[ONNX] Limit number of elements to display for list/tuple/dict in diagnostics

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106048

In a recent change, diagnostics started logging contents within tuple/list/dict
for diagnosed function arguments and return types. This brought slow down
to export due to some extremely large container instances, such as the fx to 
onnx node mapping dictionary.

This PR adds a limit to how many elements the diagnostic would record for
these types. Together with https://github.com/microsoft/onnxscript/pull/922, the performance of
export w/ diagnostics is restored and improved. As shown by pyinstrument:

GPT2 time for `fx_to_onnx_interpreter.run` 17.767s -> 1.961s
xcit_large_24_p8_224 time for `fx_to_onnx_interpreter.run` 144.729s -> 4.067s

BowenBao committed:
[ONNX] Limit number of elements to display for list/tuple/dict in diagnostics

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Limit number of elements to display for list/tuple/dict in diagnostics"


In a recent change, diagnostics started logging contents within tuple/list/dict
for diagnosed function arguments and return types. This brought slow down
to export due to some extremely large container instances, such as the fx to 
onnx node mapping dictionary.

This PR adds a limit to how many elements the diagnostic would record for
these types. Together with https://github.com/microsoft/onnxscript/pull/922, the performance of
export w/ diagnostics is improved. As shown by pyinstrument:

GPT2 time for `fx_to_onnx_interpreter.run` 17.767s -> 1.961s
xcit_large_24_p8_224 time for `fx_to_onnx_interpreter.run` 144.729s -> 4.067s

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/fx/onnxfunction_dispatcher.py:
```diff
@@ -419,7 +418,11 @@ def __init__(
+        op_schema = self.onnxfunction.op_schema
+        # Both `OnnxFunction` and `TracedOnnxFunction` never return None for `op_schema`.
+        # However their base class would. Hence return type is annotated as Optional[OpSchema].
+        assert op_schema is not None
```
cc @justinchuby for thoughts

titaiwangms replied:
What is the issue?

BowenBao replied:
Kinda unsafe assertion.

Meanwhile, reason for the change is I wasn't able to get intellisense with previous code somehow. Pylance probably still thinks `self.op_schema` is `OpSchema | None`.

justinchuby replied:
Should we continue and log? It don't expect this to fail but just in case

BowenBao replied:
It if fails it will get reported by the inflight diagnostic.
Continuing also will eventually fail since all code expect it to be not None, so better to fail early?

I wonder if there is some way to annotate correctly. Do we really need `Optional` for `op_schema` for `Op`/`OpLike`?

justinchuby replied:
Good point. Maybe we don't. We also need better opschema gen support for ops with variadic inputs I suppose. cc @gramalingam

BowenBao commented on torch/onnx/_internal/fx/onnxfunction_dispatcher.py:
```diff
@@ -498,7 +501,7 @@ def perfect_match_inputs(
                 diagnostic.with_additional_message(
-                    f"#### Failed: input type mismatch! \n"
+                    f"#### Failed: input type mismatch for input '{schema_input.name}'! \n"
                     f"Actual {torch_input_compatible_types} vs\n"
```
Unrelated question: I recall seeing `torch_input_compatible_types` as empty set when `torch_input` is `None`. Should it count as perfect match if `schema_input` is `optional` or not `required`? @titaiwangms

titaiwangms replied:
The op doesn't need an input? Could you point me to that op? Do we have optional input cases? I guess I haven't thought of this kind of case.

titaiwangms replied:
I see some of Optional[INT64]. The logic now relies on the order of the inputs. I will see if I can get the info from OpSchema.

BowenBao replied:
<img width="592" alt="image" src="https://github.com/pytorch/pytorch/assets/9376104/9769bbb4-4243-4062-88dc-ad27f48159ef">

titaiwangms replied:
I think I can add another condition saying that if Optional is allowed, this can be seen as passed. Just wanted to make sure, the optional input still needs to be presenting as "None", right?

titaiwangms replied:
https://github.com/pytorch/pytorch/issues/106055

BowenBao replied:
> I think I can add another condition saying that if Optional is allowed, this can be seen as passed. Just wanted to make sure, the optional input still needs to be presenting as "None", right?

Very likely. I'm unsure about the details, you'd need to check how dispatcher and graph building is handling this today. This op is exported correctly, so probably just need to refine logic in perfect match.

justinchuby replied:
Yes. Required Optional[...] inputs needs `None` or `Tensor`

titaiwangms commented on torch/onnx/_internal/fx/diagnostics.py:
```diff
@@ -21,6 +21,9 @@
 _LENGTH_LIMIT: int = 120
+# NOTE: The following limits are for the number of items in a list, tuple or dict.
+# The limits are set to be small to avoid heavy workload.
+_CONTAINER_ITEM_LIMIT: int = 10
```
I think we should reason that args/kwargs usually smaller than 10.

BowenBao replied:
rephrased

titaiwangms reviewed (approved):
Wait for @justinchuby to confirm the empty op_schema scenario.

BowenBao committed:
rephrase comment on "[ONNX] Limit number of elements to display for list/tuple/dict in diagnostics"


In a recent change, diagnostics started logging contents within tuple/list/dict
for diagnosed function arguments and return types. This brought slow down
to export due to some extremely large container instances, such as the fx to 
onnx node mapping dictionary.

This PR adds a limit to how many elements the diagnostic would record for
these types. Together with https://github.com/microsoft/onnxscript/pull/922, the performance of
export w/ diagnostics is restored and improved. As shown by pyinstrument:

GPT2 time for `fx_to_onnx_interpreter.run` 17.767s -> 1.961s
xcit_large_24_p8_224 time for `fx_to_onnx_interpreter.run` 144.729s -> 4.067s

[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge

justinchuby commented:
@pytorchbot merge -f "unstable rocm tests blocked"
